Python/ceval.c

```diff
@@ -2121,9 +2121,26 @@ _PyEval_UnpackIterableStackRef(PyThreadState *tstate, _PyStackRef v_stackref,
             return 1;
         }
         Py_DECREF(w);
-        _PyErr_Format(tstate, PyExc_ValueError,
-                      "too many values to unpack (expected %d)",
-                      argcnt);
+
+        if (PyList_CheckExact(v) || PyTuple_CheckExact(v)
+              || PyDict_CheckExact(v)) {
+            ll = PyDict_CheckExact(v) ? PyDict_Size(v) : Py_SIZE(v);
+            if (ll <= argcnt) {
+                _PyErr_Format(tstate, PyExc_ValueError,
+                            "too many values to unpack (expected %d)",
+                            argcnt);
+            }
+            else {
+                _PyErr_Format(tstate, PyExc_ValueError,
+                            "too many values to unpack (expected %d, got %zd)",
+                            argcnt, ll);
+            }
+        }
+        else {
+            _PyErr_Format(tstate, PyExc_ValueError,
+                          "too many values to unpack (expected %d)",
+                          argcnt);
+        }
```

Comment:
You can simplify this: put in one `(expected %d, got %zd)` and `goto Error`, and have all other cases fall back to the original.